### DISCUSSION TITLE
Thread devices get detached from OTBR when data rate is 10 sample per second

### DISCUSSION THREAD

**Discussion by nikhil1983** [openthread/openthread#8184] [Q&A]
Hello, we have developed a product based on open thread, where we are transmitting data at high data rate i.e around 10 payload per second. And each payload is of around 100 bytes. And my sleepy end device used mqtt-sn at application level, with a poll period set of 20ms. Mqtt-sn gateway and border router is setup on a RPi.  Currently for testing purpose OTBR and end devices are kept in line of sight with just 4_5 meter distance between them. 

So recently we found out that after a couple of minutes of data transfer at a high data rate (10sample per sec) **my thread devices get detached from the network** and after 5-8 seconds they again get attached to the thread network and start transmitting data over the mqtt-sn protocol. And this detached attached routine is happening once a minute. 

1. I would like to know what could be the possible reasons why my device gets detached too often ?

2. Also does thread network handle 10 samples per second data rate for a longer period of time ? Will it sustain ? 

3. And are there any api available which I can use in my firmware to force connect my thread device to a network (when in  detached mode) instead of waiting for 4_5 seconds after getting detached ?

**Comment by jwhui** [comment]
> I would like to know what could be the possible reasons why my device gets detached too often ?

There are a number of reasons why your device may be losing connection. To help diagnose the problem, can you provide more information? For example:
1. Logs from OTBR and/or sleepy end device.
2. Packet capture file for viewing in wireshark.
3. Device configuration (hw platforms used for OTBR, RCP, and Thread device)

> Also does thread network handle 10 samples per second data rate for a longer period of time ? Will it sustain ?

Thread can support far more than 10 packets per second. But note that the channel capacity is a shared resource. If you have lots of devices transmitting this much data, the channel will saturate.

> And are there any api available which I can use in my firmware to force connect my thread device to a network (when in detached mode) instead of waiting for 4_5 seconds after getting detached ?

A device will automatically start the reattach process when it detects that it becomes disconnected from the Thread network. The [minimum backoff time for reattaching is already set to ~250ms](https://github.com/openthread/openthread/blob/2cab708443ece91555a9ba6d238b5908506ac409/src/core/config/mle.h#L135), which is pretty short already.

**Reply by nikhil1983** [reply]
Okay I forgot to describe that I was using the open thread border router version provided in nRF5 SDK for Thread. Also we are using an NCP device. 

Hardware platform used for OTBR is RPi 3B+ and thread devices are using nRF52840 SoC.

**Reply by jwhui** [reply]
It sounds like you are using a relatively old version of OpenThread. Also, nRF5 SDK for Thread is now [in maintenance mode](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fstruct_sdk%2Fstruct%2Fsdk_thread_zigbee_latest.html). Is it possible for you to move to a more recent version of OpenThread?

One benefit of running RCP on OTBR is much easier access to detailed logs.

**Reply by nikhil30081995** [reply]
Hello, 
As our development is already done using nRF5 SDK for Thread we can't move to a newer version of SDK. Till now everything was working fine with upload intervals of 4seconds to 60 seconds. 

I have attached the nRF5 Sniffer capture of the issue which i am facing. My end device is configured with a child timeout period of 40 seconds and it' programmed to transfer sensor data at 1 sample/100ms. 

But after 1-2 minute max my thread device gets detached (state changes to 1). 

Below is the wire shark capture (please download from the link below) :  
https://drive.google.com/file/d/1vYAQ7La_l6tyvkWgrzTCtKzA_kG-uVre/view?usp=sharing

Can you please look at the above sniffer capture and tell what's the issue ?

**Reply by jwhui** [reply]
Can you provide the value configured for `OPENTHREAD_CONFIG_MAC_DEFAULT_MAX_FRAME_RETRIES_DIRECT`? We found that a few transmission failures is common in real world environments. As a result, we had increased the number of retries both in specification and implementation.

Can you also provide your network key so that we can better analyze the packet capture?

**Reply by nikhil30081995** [reply]
The following OpenThread commit has been used to generate pre-built Thread stack libraries, and for testing and verification: - https://github.com/openthread/openthread/commit/c6a258e3a5bd90aa26d8245c435c0ae5797027f4 

The master key for my network is : 00112233445566778899aabbccddeeff .

**Reply by jwhui** [reply]
With c6a258e3a5bd90aa26d8245c435c0ae5797027f4, the default value for `OPENTHREAD_CONFIG_MAC_DEFAULT_MAX_FRAME_RETRIES_DIRECT` is: https://github.com/openthread/openthread/blob/c6a258e3a5bd90aa26d8245c435c0ae5797027f4/src/core/config/mac.h#L73

Looking at the packet trace, there seems to be times where the number of retransmissions exceeds that value. Without additional information, it is difficult to determine the cause of those transmission failures. It could be due to external interference or the receiver is unable to generate an acknowledgement for whatever reason. I would try increasing `OPENTHREAD_CONFIG_MAC_DEFAULT_MAX_FRAME_RETRIES_DIRECT`, but that likely requires a new build of the OpenThread library.

I would reach out to Nordic to see what options they might suggest.

**Reply by nikhil30081995** [reply]
What could be ideal value for OPENTHREAD_CONFIG_MAC_DEFAULT_MAX_FRAME_RETRIES_DIRECT than ?

**Reply by jwhui** [reply]
The latest on main branch: https://github.com/openthread/openthread/blob/bacf8d625f50dd5c4b415caee754c934ba1a262c/src/core/config/mac.h#L73

All default config values are maintained here: https://github.com/openthread/openthread/tree/main/src/core/config

**Reply by nikhil30081995** [reply]
Okay i have changed the macro in the mac.h file, but how to build openthread now ?

Here is the image 

![image](https://user-images.githubusercontent.com/84308431/191174899-e5bf96f2-834c-428a-b495-a17ea5aa5a58.png)

**Reply by jwhui** [reply]
Given that you are using the nRF5 SDK for Thread, I would reach out to Nordic for further guidance.

**Reply by nikhil30081995** [reply]
> The latest on main branch:
> 
> https://github.com/openthread/openthread/blob/bacf8d625f50dd5c4b415caee754c934ba1a262c/src/core/config/mac.h#L73
> 
> All default config values are maintained here: https://github.com/openthread/openthread/tree/main/src/core/config

Hello, i have build the repo.  and replace the build libraries which were  used in my nRF5 project reffering following link : 
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsdk_tz_v4.1.0%2Fthread_ot_libraries_building.html

I would like to know, how can i verify if the value has been updated successfully or not ?

**Reply by jwhui** [reply]
After the SED attaches, you can disable its parent and observe how many retransmissions the SED performs on its data messages.

**Reply by nikhil30081995** [reply]
`96d1dea311f otbr-agent[216]: 00:01:12.373 [I] MeshForwarder-:     src:[fd68:4f54:dead:1:52de:cf79:cf91:8abf]
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.373 [I] MeshForwarder-:     dst:[ff33:40:fd88:f3df:b78f:478b:0:1]
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.511 [I] Mle-----------: Send Child Update Request to child (fe80:0:0:0:58fc:21e3:10df:82ad,0x3801)
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.511 [I] MeshForwarder-: Prepping indir tx IPv6 UDP msg, len:172, chksum:f7c7, ecn:no, to:0x3801, sec:no, prio:net
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.555 [I] DataPollHandlr: Rx data poll, src:0x3801, qed_msgs:1, rss:-59, ack-fp:1
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.558 [I] Mac-----------: Frame rx failed, error:Duplicated, len:22, seqnum:204, type:Cmd(DataReq), src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.568 [I] Mac-----------: Frame tx attempt 1/1 failed, error:NoAck, len:126, seqnum:164, type:Data, src:3ea40ae696963b03, dst:5afc21e310df82ad, sec:yes, ackreq:yes
Sep 22 08:29:13 096d1dea311f otbr-agent[216]: 00:01:12.568 [I] DataPollHandlr: Indirect tx to child 3801 failed, attempt 1/4
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.667 [I] DataPollHandlr: Rx data poll, src:0x3801, qed_msgs:1, rss:-64, ack-fp:1
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.674 [I] Mac-----------: Frame tx attempt 1/1 failed, error:NoAck, len:126, seqnum:164, type:Data, src:3ea40ae696963b03, dst:5afc21e310df82ad, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.674 [I] DataPollHandlr: Indirect tx to child 3801 failed, attempt 2/4
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.680 [I] Mac-----------: Frame rx failed, error:Duplicated, len:22, seqnum:205, type:Cmd(DataReq), src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.686 [I] Mac-----------: Frame rx failed, error:Duplicated, len:22, seqnum:205, type:Cmd(DataReq), src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.777 [I] MeshForwarder-: Received IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, from:0x3801, sec:yes, prio:normal, rss:-64.0
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.777 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:2ea4:b9e4:2714:ee]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.777 [I] MeshForwarder-:     dst:[ff03:0:0:0:0:0:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.777 [I] Platform------: MulticastRoutingManager: AddMulticastForwardingCache: add dynamic route: Thread fd88:f3df:b78f:478b:2ea4:b9e4:2714:ee => ff03:0:0:0:0:0:0:1 None: OK
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.777 [I] Platform------: MulticastRoutingManager: ProcessMulticastRouterMessages: OK
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.782 [I] Mac-----------: Frame rx failed, error:Duplicated, len:53, seqnum:206, type:Data, src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.786 [I] Mac-----------: Frame rx failed, error:Duplicated, len:53, seqnum:206, type:Data, src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.793 [I] MeshForwarder-: Sent IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, to:0xffff, sec:yes, prio:low
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.793 [I] MeshForwarder-:     src:[fd68:4f54:dead:1:52de:cf79:cf91:8abf]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.793 [I] MeshForwarder-:     dst:[ff33:40:fd88:f3df:b78f:478b:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.796 [I] Mac-----------: Frame rx failed, error:Duplicated, len:53, seqnum:206, type:Data, src:0x3801, dst:0x3800, sec:yes, ackreq:yes
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.813 [I] MeshForwarder-: Sent IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, to:0xffff, sec:yes, prio:low
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.813 [I] MeshForwarder-:     src:[fd68:4f54:dead:1:52de:cf79:cf91:8abf]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.813 [I] MeshForwarder-:     dst:[ff33:40:fd88:f3df:b78f:478b:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.817 [I] MeshForwarder-: Sent IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, to:0xffff, sec:yes, prio:normal
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.817 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:2ea4:b9e4:2714:ee]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.817 [I] MeshForwarder-:     dst:[ff03:0:0:0:0:0:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.828 [I] MeshForwarder-: Received IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, from:0x5800, sec:yes, prio:normal, rss:-46.0
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.828 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:401e:9a87:506e:9392]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.828 [I] MeshForwarder-:     dst:[ff33:40:fdde:ad00:beef:0:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.838 [I] MeshForwarder-: Received IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, from:0x5800, sec:yes, prio:normal, rss:-46.0
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.838 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:2ea4:b9e4:2714:ee]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.838 [I] MeshForwarder-:     dst:[ff03:0:0:0:0:0:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.841 [I] MeshForwarder-: Received IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, from:0xc800, sec:yes, prio:normal, rss:-51.0
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.841 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:4b5c:caac:de38:da73]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.841 [I] MeshForwarder-:     dst:[ff33:40:fdde:ad00:beef:0:0:1]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.845 [I] MeshForwarder-: Sent IPv6 HopOpts msg, len:59, chksum:0000, ecn:no, to:0xffff, sec:yes, prio:normal
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.845 [I] MeshForwarder-:     src:[fd88:f3df:b78f:478b:2ea4:b9e4:2714:ee]
Sep 22 08:29:14 096d1dea311f otbr-agent[216]: 00:01:12.845 [I] MeshForwarder-`


Here are the logs from OTBR.
